Throw error in case of duplicate output file request from yaml

ekdejong opened the pull request:
Addresses [ issue #7014 ](https://github.com/E3SM-Project/E3SM/issues/7014) by checking for a duplicate *.nc file that would be created by a yaml using the same prefix, averaging type, and interval.

Commits have been tested for expected behavior on LC Intel (Dane) with an older version of E3SM.

Fixes #7014

[BFB]

AaronDonahue commented:
@bartgol , Emily is a PostDoc working with us at LLNL and she was hitting an error that matched this issue from almost a year ago.  She came up with a fix so I encouraged her to submit a PR.

AaronDonahue reviewed (approved):
Looks good to me.

bartgol reviewed (approved):
This looks good. At some point, we should prob move some checks to the buildnml scripts, so that we can catch these before submission. But this is def an improvement.

rljacob commented:
Thanks for the PR @ekdejong .  Just update the title to be a "verb noun" sentence about the PR instead of the branch name.

bartgol commented:
@ekdejong The v1 tests fail the RUN phase, so something may be amiss with the logic you implemented.
```

BACKTRACE:
/home/runner/_work/E3SM/E3SM/components/eamxx/src/share/io/eamxx_output_manager.cpp:432
Error! Output file already exists: 
  - filename: ERS.ne4pg2_ne4pg2.F2010-SCREAMv1.ghci-snl-cpu_gnu.eamxx-prod.C.20251125_214248_yvz7p8.scream.1hourlyAVG_native.h.AVERAGE.nhours_x1.1994-10-07-00000.nc
  - timestamp: 1994-10-07-03600
This likely indicates that two yaml files were specified with identical interval, average type, and prefix.

FAILED CONDITION: 'false'
```

Did you test your fix with restarts? The failing tests are all restart tests. The ERS tests that passed do not produce any eamxx output file...

mahf708 reviewed (changes requested):
If the goal is to prevent duplicating yaml files, then this check should be impl'ed in the buildnml python stuff.

If the goal is to prevent duplicating any output files that will overwrite each other, that's a pretty dangerous and will require more discussion. Consider the following situation: a simulation fails at 2010-05-01 and the prior restart was 2010-01-01 with monthly averages for 01, 02, 03, and 04 written; if we restart from 01-01, then this check won't allow us to proceed because the monthly files already exist, but our intention is to overwrite those files with newer data.

ekdejong commented:
@mahf708 and @bartgol I've added an additional check in the eamxx namelist file python setup as suggested by @mahf708 -- you'll have to tell me if it's a "nice or ugly way" ;P 

I also added some additional logic to the original output manager hack that should allow for a useful error message but not interfere with restarts. In particular, there is now an option to add "allow_overwrite_existing_files: true" to one's yaml file if the user truly intends to overwrite existing output; else the flag will trigger a meaningful EKAT error, including a tip regarding that yaml option. However, if all of this is too complex, I can remove those changes to the output manager and we can just roll with the namelist python fix for now.

mahf708 commented:
oh btw, I triggered an AI review out of curiosity what it would say ;) and lol, it is mostly useless (love how it is insisting on uppercase YAML...)

bartgol commented on components/eamxx/src/share/io/eamxx_output_manager.cpp:
```diff
@@ -426,6 +426,20 @@ void OutputManager::run(const util::TimeStamp& timestamp)
+            "  - timestamp: " + timestamp.to_string() + "\n"
+ 	    "Check that you do not have a duplicated yaml file,\n"
+            "or if you intend to overwrite, set allow_overwrite_existing_files: true");
+      }    
```
This is a good point. I think you can simply add a check in any of the io tests (maybe io_basic?). E.g., you could change

```cpp
 // Create Output manager
  OutputManager om; 

  // Attempt to use invalid fp precision string
  om_pl.set("floating_point_precision",std::string("triple"));
  om.initialize(comm,om_pl,t0,false);
  REQUIRE_THROWS (om.setup(fm,gm->get_grid_names()));
  om_pl.set("floating_point_precision",std::string("single"));
  om.initialize(comm,om_pl,t0,false);
  om.setup(fm,gm->get_grid_names());
```
to
```cpp
  // Create Output manager
  auto create_om = [&](const ekat::ParameterList& pl) {
    OutputManager om;
    om.initialize(comm,pl,t0,false);
    om.setup(fm,gm->get_grid_names());
    return om;
  };  
  
  om_pl.set("floating_point_precision",std::string("triple"));
  REQUIRE_THROWS(create_om(om_pl)); // bad precision
  om_pl.set("floating_point_precision",std::string("single"));
  auto om = create_om(om_pl);
  REQUIRE_THROWS (create_om(om_pl)); // Cannot create 2 om's with same specs!
```

ekdejong commented:
@mahf708 @bartgol I've opted to take Naser's advice here -- I've updated the python code addition to use the proper `expect` syntax, but abandoned the C++ additions. I'll let the "IO epicists" take that one on instead ;)

mahf708 reviewed (approved):
🚀

bartgol reviewed (approved):
Thanks!
